File: api/internal/training/training.go
// Package training answers GET /api/training.
//
// This is the endpoint where invariant 2 becomes visible from outside. A track
// has no state column; its state is counted in v_track_states from the systems
// that prove it (ADR 0003), and nothing in this package decides one. The whole
// job here is assembly: three flat answers from the store become the tree the
// contract describes, in the order the queries returned them.
//
// Two rules worth stating in Go, because Go is where they are easiest to break:
//
// A track with no evidence keeps its place. Nine of the twenty-two tracks are
// self-study at launch, and dropping them — or letting an inner join drop them —
// would make the log look fuller than it is. They come back with an empty array
// and a note that says so, which the interface renders as
// `NO SYSTEM YET → SELF-STUDY`.
//
// The two header numbers are counted from the rows that are actually served, not
// asked of the database a second time. A second count is a second truth, and the
// one place it can differ is the one place it matters: the line above the list
// would disagree with the list.
//
// The handler carries the shape httpx.StrictServerInterface expects even though
// the router mounts it directly (ADR 0016). When the generated router takes over
// in C7 that has to be a router change and not a rewrite.
package training

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"time"

	"github.com/G1NG4R/timseil-dev/api/internal/httpx"
	"github.com/G1NG4R/timseil-dev/api/internal/store"
)

// cacheControl is what the contract declares for this path
// (components/headers/CacheControlMedium). The value lives in httpx, held
// against the served document there; this line only picks which of the four
// directives this path carries, and the contract test next door checks that the
// pick is the right one (ADR 0009).
const cacheControl = httpx.CacheControlMedium

// selfStudyNote is the API side of `NO SYSTEM YET → SELF-STUDY`.
//
// A constant rather than a column: the contract sets the field only when
// `evidence` is empty, so a column would carry the same nine words nine times
// and a migration to hold them. Lowercase because the API is lowercase and the
// interface uppercases (handbook ch. 14).
const selfStudyNote = "self-study"

// Queries is the slice of the store this endpoint needs. Narrow on purpose: it
// is what lets every branch below — an empty database, a track without
// evidence, a broken query — be tested without Postgres.
type Queries interface {
	ListModules(ctx context.Context) ([]store.ListModulesRow, error)
	ListTracksWithState(ctx context.Context) ([]store.ListTracksWithStateRow, error)
	ListTrackEvidence(ctx context.Context) ([]store.ListTrackEvidenceRow, error)
}

type Handler struct {
	queries Queries
	log     *slog.Logger

	// now is injected so the tests can assert on generatedAt without racing it.
	now func() time.Time
}

func New(q Queries, log *slog.Logger) *Handler {
	return &Handler{queries: q, log: log, now: time.Now}
}

// GetTraining carries the signature of httpx.StrictServerInterface.
//
// There is no honest partial answer here. A training log without its evidence
// lines is not a degraded version of this document — it is the claim that
// twenty-two tracks rest on nothing, which is a different claim and a false one.
// So any query error is an error, and the adapter turns it into a problem
// document.
func (h *Handler) GetTraining(ctx context.Context, req httpx.GetTrainingRequestObject) (
	httpx.GetTrainingResponseObject, error,
) {
	moduleRows, err := h.queries.ListModules(ctx)
	if err != nil {
		return nil, err
	}
	trackRows, err := h.queries.ListTracksWithState(ctx)
	if err != nil {
		return nil, err
	}
	evidenceRows, err := h.queries.ListTrackEvidence(ctx)
	if err != nil {
		return nil, err
	}

	body := assemble(moduleRows, trackRows, evidenceRows)

	etag, err := etagOf(body)
	if err != nil {
		return nil, err
	}

	if httpx.MatchesETag(ifNoneMatch(req), etag) {
		return httpx.GetTraining304Response{
			Headers: httpx.NotModifiedResponseHeaders{ETag: &etag},
		}, nil
	}

	// Filled after the tag is computed, deliberately — see etagOf.
	body.GeneratedAt = h.now().UTC()

	cache := cacheControl
	return httpx.GetTraining200JSONResponse{
		Body: body,
		Headers: httpx.GetTraining200ResponseHeaders{
			CacheControl: &cache,
			ETag:         &etag,
		},
	}, nil
}

// ------------------------------------------------------------------ assembly

// assemble turns the three flat answers into the contract's tree.
//
// Track ids are the join key and they stop here: the contract exposes none, so
// no caller ever learns one. Everything else is a copy, in the order the queries
// produced — module_no, then sort_order, then system_no inside a track.
//
// The grouping map is read through the ordered slices and never ranged over. Map
// iteration in Go is deliberately random, and a response whose order changes
// between two identical database states has a different ETag every time, which
// would leave the 304 path dead without anything looking broken.
func assemble(
	moduleRows []store.ListModulesRow,
	trackRows []store.ListTracksWithStateRow,
	evidenceRows []store.ListTrackEvidenceRow,
) httpx.Training {
	evidence := make(map[int64][]httpx.Evidence, len(evidenceRows))
	systems := make(map[string]struct{}, len(evidenceRows))

	for _, row := range evidenceRows {
		evidence[row.TrackID] = append(evidence[row.TrackID], httpx.Evidence{
			SystemId: row.Slug,
			SystemNo: row.SystemNo,
			Detail:   row.Detail,
		})
		systems[row.Slug] = struct{}{}
	}

	tracks := make(map[string][]httpx.Track, len(moduleRows))
	for _, row := range trackRows {
		track := httpx.Track{
			Name:  row.Name,
			State: httpx.TrackState(row.State),
			// Never nil: the contract requires the array, and `null` where an
			// array is required is a shape no generated client expects.
			Evidence: []httpx.Evidence{},
		}
		if lines, ok := evidence[row.TrackID]; ok {
			track.Evidence = lines
		} else {
			note := selfStudyNote
			track.Note = &note
		}
		tracks[row.ModuleNo] = append(tracks[row.ModuleNo], track)
	}

	// An empty database is an answer, not a failure: `modules: []` with two
	// zeroes is what this endpoint says before the seed has ever run.
	modules := make([]httpx.Module, 0, len(moduleRows))
	trackCount := 0
	for _, row := range moduleRows {
		module := httpx.Module{
			No:     row.ModuleNo,
			Title:  row.Title,
			Tracks: []httpx.Track{},
		}
		if list, ok := tracks[row.ModuleNo]; ok {
			module.Tracks = list
		}
		trackCount += len(module.Tracks)
		modules = append(modules, module)
	}

	return httpx.Training{
		Modules:    modules,
		TrackCount: trackCount,
		// Distinct systems behind the lines this document actually carries. On
		// launch day every one of the thirteen points at 02 timseil.dev, so the
		// header reads one system rather than rounding the log up to thirteen.
		EvidenceSystems: len(systems),
	}
}

// ---------------------------------------------------------------- the adapter

// ServeHTTP adapts the strict signature to a route.
//
// It is what the generated router replaces in C7. There are no sentinel errors
// in this package because this path declares no 400 and no 404: it takes no
// parameter that can be wrong, and the training log exists as soon as the
// service does.
func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	var params httpx.GetTrainingParams
	if match := r.Header.Get("If-None-Match"); match != "" {
		params.IfNoneMatch = &match
	}

	resp, err := h.GetTraining(r.Context(), httpx.GetTrainingRequestObject{Params: params})
	if err != nil {
		httpx.WriteInternalProblem(w, r, h.log, err)
		return
	}
	if err := resp.VisitGetTrainingResponse(w); err != nil {
		// The status line is already out; there is nothing left to send but a
		// log line.
		h.log.Error("writing the training response", "err", err)
	}
}

// etagOf hashes the answer without generatedAt.
//
// With it in the hash the tag would change every time the clock does,
// If-None-Match would never match, and the 304 path would be dead code that
// nobody notices — the endpoint would still look correct and would ship a full
// body to every poll.
//
// The body is taken by value, so blanking the field is local to the hash and the
// caller's copy keeps the real time.
func etagOf(body httpx.Training) (string, error) {
	body.GeneratedAt = time.Time{}
	encoded, err := json.Marshal(body)
	if err != nil {
		return "", err
	}
	return httpx.ETagOf(encoded), nil
}

func ifNoneMatch(req httpx.GetTrainingRequestObject) string {
	if req.Params.IfNoneMatch == nil {
		return ""
	}
	return *req.Params.IfNoneMatch
}

var _ http.Handler = (*Handler)(nil)

File: web/lib/api/training.ts
// What the homepage reads off one `/api/training` answer.
//
// IT IS HERE AND NOT IN lib/home/ FOR lib/api/systems.ts's REASON, and the two
// files are deliberately the same shape: one endpoint, one file, every
// judgement about what its numbers MEAN, and nothing that imports `next/*` — so
// `node --test` can reach all of it. lib/home/sections.ts is the other kind of
// file, an order this page fixes for itself; nothing in here is ours to fix.
//
// THIS ENDPOINT IS WHERE INVARIANT 2 BECOMES VISIBLE. A track has no state
// column: `v_track_states` counts the systems that prove it, and ADR 0018 says
// what that leaves for this side of the wire — "H4 rendert den Baum, wie er
// kommt: Reihenfolge, Zustände und Belegzeilen sind serverseitig entschieden,
// die Oberfläche schreibt sie groß und sortiert nicht nach." So there is no
// sort in this file, no count that the answer already carries, and no branch
// that decides a state. What is here is spelling.
//
// THE EMPTY ROW IS THE NORMAL ONE. Nine of the twenty-two tracks have no
// evidence at all at launch, and they are the reason the endpoint exists in
// this shape — a log that dropped them would look fuller than the systems
// behind it. Every function below is written for that row first.
//
// EVERY FIELD IS READ DEFENSIVELY THOUGH THE TYPE IS GENERATED, the rule
// lib/api/values.ts carries: the generated type describes the contract, not the
// bytes, and ADR 0035's overlapping start puts a real case behind the
// difference.

import { NO_DATA, TRACK_MARKS, isTrackState, type TrackState } from "../state/words.ts";

import type { GetBody } from "./client.ts";
import { finiteNumber, nonEmpty } from "./values.ts";

/** The whole log, as the contract answers it. */
export type Training = GetBody<"/api/training">;

/**
 * The line under a track's name: a prefix, and what it points at.
 *
 * TWO FIELDS AND NOT ONE STRING, because the sheet colours them apart — the
 * prefix is dim, the systems are bright, and on hover only the second half
 * lifts. A component that received `"SHIPPED IN → 02 TIMSEIL-DEV"` would have to
 * split it again to draw it.
 */
export interface EvidenceLine {
  /** `RUNS IN`, `SHIPPED IN`, `RUNNING IN`, `PLANNED IN`, `NO SYSTEM YET`. */
  readonly prefix: string;
  /**
   * The systems, or the note that stands in for them, or `null`.
   *
   * `null` IS NOT AN ERROR HERE. `NO SYSTEM YET` is already a complete
   * sentence; when the answer carries no note either, the row says that much
   * and stops rather than printing an arrow into nothing.
   */
  readonly text: string | null;
}

/** One track: a name, what it is in, and what proves it. */
export interface TrackView {
  readonly name: string;
  /**
   * `null` when the answer carried a state this build does not know.
   *
   * INVARIANT 1 RATHER THAN A GUESS. A word from a contract newer than this
   * container is not a track state we may render, and mapping it onto the
   * nearest one we do know would be an invented claim about somebody's skill.
   * The row keeps its name, says `— NO DATA` where the word goes, and draws no
   * bar. ADR 0035's overlapping start is when this happens.
   */
  readonly state: TrackState | null;
  /** How many of the four bar segments are filled. `0` for an unknown state. */
  readonly steps: number;
  readonly evidence: EvidenceLine;
}

/** One module card: its number, its name, and its tracks in the server's order. */
export interface ModuleView {
  readonly no: string;
  readonly title: string;
  readonly tracks: readonly TrackView[];
}

/**
 * The prefix each state gives its evidence line.
 *
 * FIVE PREFIXES FOR FOUR STATES, and the fifth is not a state at all: an empty
 * `evidence` array wins over whatever the track is in, because "nothing to
 * point at" is the more specific fact. Today that is nine of twenty-two rows.
 *
 * THE SHEET WRITES THE `learning` CASE TWICE — `TOUCHED IN → 02 RELAY (TOKEN
 * SIGNING)` in one row and `RUNNING IN → 04 TIMSEIL.DEV (UPTIME MONITOR)` in
 * another, for the same state. One of them had to go: two spellings of one
 * prefix are two rows that look like different kinds of evidence and are not.
 * `RUNNING IN` is the one kept — `TOUCHED IN` describes how much of the system
 * the track accounts for, which is a judgement no column carries. Recorded as a
 * sheet finding rather than settled quietly.
 */
const PREFIX: Record<TrackState, string> = {
  core: "RUNS IN",
  applied: "SHIPPED IN",
  learning: "RUNNING IN",
  queued: "PLANNED IN",
};

/** What the line says when the array is empty, whatever the state. */
const NO_EVIDENCE = "NO SYSTEM YET";

/**
 * The prefix for a row that has systems but a state this build cannot name.
 *
 * Neutral on purpose. The four in `PREFIX` each say what the evidence MAKES the
 * track, and that is the half a newer contract has taken away; this one says
 * only what the rows are. `NO SYSTEM YET` would be plainly false about a row
 * that names two.
 */
const UNKNOWN_PREFIX = "EVIDENCE";

/**
 * One evidence row, spelled the way the sheet spells it.
 *
 * `02 TIMSEIL-DEV (API, HEALTH ENDPOINT)` — the system's number, its id, and
 * the detail in brackets. Uppercased here rather than in CSS because the api
 * speaks lowercase and the interface shouts (handbook ch. 14), and because a
 * `text-transform` cannot be read back by a test that asks what the row says.
 *
 * THE SHEET WRITES `TIMSEIL.DEV` AND THIS WRITES `TIMSEIL-DEV`, which is a
 * one-character divergence with a reason. `Evidence` carries `systemId` — the
 * slug — and no display name; the slug is what `/work/timseil-dev` uses and
 * what every other surface on this site shows. Inventing the dot would mean
 * mapping slugs to names in the browser off a second endpoint, for punctuation.
 */
function evidenceRow(row: unknown): string | null {
  const raw = (row ?? {}) as Record<string, unknown>;
  const no = nonEmpty(raw.systemNo);
  const id = nonEmpty(raw.systemId);
  if (id === null) return null;

  const head = no === null ? id.toUpperCase() : `${no} ${id.toUpperCase()}`;
  const detail = nonEmpty(raw.detail);
  return detail === null ? head : `${head} (${detail.toUpperCase()})`;
}

/**
 * The evidence line for one track.
 *
 * The middle dot joins several systems, which is the sheet's separator and the
 * one this site already uses in the meta bar. With one system it never appears;
 * with six it is the whole line — `RUNS IN → 02 RELAY · 03 FEEDHOUND · 04
 * TIMSEIL-DEV` — and the day that happens the mobile artboard wants it
 * abbreviated. That is not this phase's problem and it is written down.
 */
export function evidenceLine(track: unknown): EvidenceLine {
  const raw = (track ?? {}) as Record<string, unknown>;
  const rows = Array.isArray(raw.evidence) ? raw.evidence : [];
  const state = isTrackState(raw.state) ? raw.state : null;

  const systems = rows.map(evidenceRow).filter((row): row is string => row !== null);

  if (systems.length === 0) {
    // The note is the api's word for what stands in place of a system, and the
    // contract sets it only in this branch. Absent, the prefix stands alone.
    const note = nonEmpty(raw.note);
    return { prefix: NO_EVIDENCE, text: note === null ? null : note.toUpperCase() };
  }

  return { prefix: state === null ? UNKNOWN_PREFIX : PREFIX[state], text: systems.join(" · ") };
}

/** One track, read defensively. `null` when the row has no name to show. */
export function trackView(track: unknown): TrackView | null {
  const raw = (track ?? {}) as Record<string, unknown>;
  const name = nonEmpty(raw.name);
  if (name === null) return null;

  const state = isTrackState(raw.state) ? raw.state : null;

  return {
    name,
    state,
    steps: state === null ? 0 : TRACK_MARKS[state].steps,
    evidence: evidenceLine(track),
  };
}

/**
 * The five module cards, in the order the answer put them in.
 *
 * NO SORT, AND THAT IS D5 OF THIS PHASE. `ORDER BY module_no, sort_order` is
 * part of the answer rather than a convenience (ADR 0018) — it is also what
 * keeps the ETag stable. The sheet draws the cards `01 · 02 · 04 · 03 · 05`
 * because DevOps has six tracks and that balances the row; on this page the
 * number is the order, the same rule HOME.01 states for the sections above.
 */
export function modules(body: Training | null): readonly ModuleView[] {
  const raw = (body ?? {}) as unknown as Record<string, unknown>;
  const list = Array.isArray(raw.modules) ? raw.modules : [];

  const views: ModuleView[] = [];
  for (const entry of list) {
    // Not named `module`: Next refuses that identifier in application code, and
    // the rule is older than this file (@next/next/no-assign-module-variable).
    const card = (entry ?? {}) as Record<string, unknown>;
    const no = nonEmpty(card.no);
    const title = nonEmpty(card.title);
    if (no === null || title === null) continue;

    const tracks = (Array.isArray(card.tracks) ? card.tracks : [])
      .map(trackView)
      .filter((track): track is TrackView => track !== null);

    // A module with no tracks keeps its card. ADR 0018 kept it in the answer
    // for the same reason: "das Modul ist leer" is a different statement from
    // "das Modul gibt es nicht", and only one of them is true here.
    views.push({ no, title, tracks });
  }

  return views;
}

/**
 * The line over the log: what it is, how big it is, and where it comes from.
 *
 * `SELF-TRACKED · 22 TRACKS · EVIDENCE: 01 SYSTEM · SOURCE: /api/training`
 *
 * BOTH NUMBERS ARE THE ANSWER'S, NOT A COUNT TAKEN HERE. ADR 0018 decided that
 * on the other side of the wire and gave the reason this side has to honour:
 * the api counts the distinct systems of exactly the rows it serves, so "the
 * line above the list cannot disagree with the list". Counting the tree again
 * here would rebuild the second source of truth that argument removed — and it
 * would put the disagreement in the one place it shows.
 *
 * `UPDATED [DATE]` IS NOT IN THIS LINE, and the sheet draws it. The only
 * timestamp the contract carries is `generatedAt`, which
 * api/internal/training/training.go fills AFTER computing the ETag: it is the
 * moment this answer was assembled, not the day the log last changed. Printed
 * as UPDATED it would move on every reload while nothing had moved, which is
 * the class of number this whole site is an argument against. The sheet means
 * the `updated` field of its own `progress.json` draft, and no column holds it.
 *
 * The words are nomenclature and stay English — the same call `lib/home/
 * sections.ts` makes for `SYS.02 SELECTED WORK`, and the reason no count on
 * this page has ever been a dictionary key.
 */
export function trainingMeta(body: Training | null): string {
  const raw = (body ?? {}) as unknown as Record<string, unknown>;

  // `finiteNumber` and not a guard written here: lib/api/values.ts exists
  // because two copies of a guard this small are how the copies start
  // disagreeing, and the day somebody tightens it for the case study is the day
  // this one would keep the old rule.
  const tracks = finiteNumber(raw.trackCount);
  const systems = finiteNumber(raw.evidenceSystems);

  const parts = ["SELF-TRACKED"];
  parts.push(tracks === null ? `${NO_DATA} TRACKS` : `${String(tracks)} TRACKS`);
  parts.push(
    systems === null
      ? `EVIDENCE: ${NO_DATA}`
      : `EVIDENCE: ${pad(systems)} ${systems === 1 ? "SYSTEM" : "SYSTEMS"}`,
  );
  parts.push("SOURCE: /api/training");

  return parts.join(" · ");
}

/**
 * Two digits, the way every identifier on this site is written.
 *
 * `SYS.01`, `02 TIMSEIL-DEV`, `01 SYSTEM` — the padding is the house style and
 * the sheet uses it in the same line. It stops at two on purpose: a hundredth
 * system is a nicer problem than a mis-aligned column.
 */
function pad(value: number): string {
  return value < 10 && value >= 0 ? `0${String(value)}` : String(value);
}
